src/cs/Ssh/Services/AuthenticationService.cs

```diff
@@ -273,89 +363,128 @@ await Session.SendMessageAsync(new AuthenticationSuccessMessage(), cancellation)
+							}
+
+							privateKey = await privateKeyProvider(privateKey, cancellation)
+								.ConfigureAwait(false);
+						}
+
+						if (privateKey != null)
+						{
+							await RequestAuthenticationAsync(username, privateKey, cancellation)
+								.ConfigureAwait(false);
+						}
+						else
+						{
+							await Session.CloseAsync(SshDisconnectReason.AuthCancelledByUser)
+								.ConfigureAwait(false);
+						}
+					}));
+			}
 		}
 
-		var passwordCredentialProvider = credentials.PasswordProvider;
-		if (passwordCredentialProvider != null)
+		if (configuredMethods.Contains(AuthenticationMethods.Password))
 		{
-			this.clientAuthenticationMethods.Enqueue(async (cancellation) =>
+			var passwordCredentialProvider = credentials.PasswordProvider;
+			if (passwordCredentialProvider != null)
 			{
-				var passwordCredentialTask = passwordCredentialProvider(cancellation);
-				var passwordCredential = passwordCredentialTask != null
-					? await passwordCredentialTask.ConfigureAwait(false) : null;
-				if (passwordCredential != null)
-				{
-					await RequestAuthenticationAsync(
-						passwordCredential.Value.Item1 ?? string.Empty,
-						passwordCredential.Value.Item2,
-						cancellation).ConfigureAwait(false);
-				}
-				else
-				{
-					await Session.CloseAsync(SshDisconnectReason.AuthCancelledByUser)
-						.ConfigureAwait(false);
-				}
-			});
-		}
-		else if (credentials.Password != null)
-		{
-			var username = credentials.Username ?? string.Empty;
-			var password = credentials.Password;
-			this.clientAuthenticationMethods.Enqueue(async (cancellation) =>
+				this.clientAuthenticationMethods.Enqueue(
+					(AuthenticationMethods.Password, async (cancellation) =>
+					{
+						var passwordCredentialTask = passwordCredentialProvider(cancellation);
+						var passwordCredential = passwordCredentialTask != null
+							? await passwordCredentialTask.ConfigureAwait(false) : null;
+						if (passwordCredential != null)
+						{
+							await RequestAuthenticationAsync(
+								passwordCredential.Value.Item1 ?? string.Empty,
+								passwordCredential.Value.Item2,
+								cancellation).ConfigureAwait(false);
+						}
+						else
+						{
+							await Session.CloseAsync(SshDisconnectReason.AuthCancelledByUser)
+								.ConfigureAwait(false);
+						}
+					}));
+			}
+			else if (credentials.Password != null)
 			{
-				await RequestAuthenticationAsync(username, password, cancellation)
-					.ConfigureAwait(false);
-			});
+				var username = credentials.Username ?? string.Empty;
+				var password = credentials.Password;
+				this.clientAuthenticationMethods.Enqueue(
+					(AuthenticationMethods.Password, async (cancellation) =>
+					{
+						await RequestAuthenticationAsync(username, password, cancellation)
+							.ConfigureAwait(false);
+					}));
+			}
 		}
 
+		// Only add None or Interactive methods if no client credentials were supplied.
```

Comment:
I did not understand this comment. Do we try password authentication before keyboard interactive?
If the server supports both keyboard interactive and password the standard ssh client will try password authentication before keyboard interactive?

Reply:
This code runs on the client side. The client application may supply public-key or password `SshClientCredentials` ahead of time, before connecting/authenticating the session; in that case only the supplied credentials are used and there is no fallback to other methods. Otherwise the client will fallback to `none` and/or `keyboard-interactive` methods, if the server indicates support for those. It currently does not ever fallback to password auth, though maybe it could.